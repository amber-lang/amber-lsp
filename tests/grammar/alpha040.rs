use std::fs::read_to_string;

use chumsky::error::Rich;
use insta::assert_debug_snapshot;

use amber_lsp::grammar::{
    alpha040::{lexer::Token, AmberCompiler, GlobalStatement, Spanned},
    LSPAnalysis, ParserResponse,
};

fn tokenize(input: &str) -> Vec<Spanned<Token>> {
    AmberCompiler::new().tokenize(input)
}

fn parse(
    tokens: &[Spanned<Token>],
) -> (Option<Vec<Spanned<GlobalStatement>>>, Vec<Rich<'_, String>>) {
    let ParserResponse {
        ast,
        errors,
        semantic_tokens: _,
    } = AmberCompiler::new().parse(tokens);

    let ast = match ast {
        amber_lsp::grammar::Grammar::Alpha040(ast) => ast,
        _ => panic!("Unexpected AST"),
    };

    (ast, errors)
}

fn parse_unwrap(tokens: &[Spanned<Token>]) -> Vec<Spanned<GlobalStatement>> {
    let (ast, errors) = parse(tokens);
    if !errors.is_empty() {
        panic!("Errors: {:?}", errors);
    }
    ast.unwrap()
}

#[test]
fn test_stdlib_array() {
    let stdlib = read_to_string("resources/alpha040/std/array.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_date() {
    let stdlib = read_to_string("resources/alpha040/std/date.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_env() {
    let stdlib = read_to_string("resources/alpha040/std/env.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_fs() {
    let stdlib = read_to_string("resources/alpha040/std/fs.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_http() {
    let stdlib = read_to_string("resources/alpha040/std/http.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_math() {
    let stdlib = read_to_string("resources/alpha040/std/math.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_text() {
    let stdlib = read_to_string("resources/alpha040/std/text.ab").unwrap();

    let tokens = tokenize(&stdlib);

    assert_debug_snapshot!(tokens);

    assert_debug_snapshot!(parse_unwrap(&tokens));
}

#[test]
fn test_unfinished_function_call() {
    let input = "
    import { array_contains } from \"std/array\"
    
    let x = [1, 2, 3]
    let y = 2
    
    echo array_contains(x, y)
    
    let line = 213
    
    for idx, line in lines(\"text.txt\") {
      echo line
    }
    
    // fun foo(x: Num, y: Text) {
    
    // }
    
    fun foo(x: Num, y: Text) {
    }
    
    // fun abc() {}
    
    array_contains([1, 2, 3],)
    ";

    let tokens = tokenize(input);
    assert_debug_snapshot!(tokens);

    let result = parse(&tokens);

    assert_debug_snapshot!(result);
}
