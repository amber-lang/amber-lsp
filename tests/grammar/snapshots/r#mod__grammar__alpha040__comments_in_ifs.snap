---
source: tests/grammar/alpha040.rs
expression: parse(&tokenize(input))
---
(
    Some(
        [
            (
                Statement(
                    (
                        IfChain(
                            (
                                "if",
                                5..7,
                            ),
                            [
                                (
                                    IfCondition(
                                        (
                                            InlineIfCondition(
                                                (
                                                    Eq(
                                                        (
                                                            Number(
                                                                (
                                                                    1.0,
                                                                    18..19,
                                                                ),
                                                            ),
                                                            18..19,
                                                        ),
                                                        (
                                                            Number(
                                                                (
                                                                    2.0,
                                                                    23..24,
                                                                ),
                                                            ),
                                                            23..24,
                                                        ),
                                                    ),
                                                    18..24,
                                                ),
                                                (
                                                    Echo(
                                                        (
                                                            "echo",
                                                            26..30,
                                                        ),
                                                        (
                                                            Text(
                                                                [
                                                                    (
                                                                        Text(
                                                                            (
                                                                                "x",
                                                                                32..33,
                                                                            ),
                                                                        ),
                                                                        32..33,
                                                                    ),
                                                                ],
                                                            ),
                                                            31..34,
                                                        ),
                                                    ),
                                                    26..34,
                                                ),
                                            ),
                                            18..34,
                                        ),
                                    ),
                                    18..34,
                                ),
                                (
                                    IfCondition(
                                        (
                                            Comment(
                                                (
                                                    Comment(
                                                        "test comment",
                                                    ),
                                                    35..51,
                                                ),
                                            ),
                                            35..51,
                                        ),
                                    ),
                                    35..51,
                                ),
                                (
                                    IfCondition(
                                        (
                                            Comment(
                                                (
                                                    Comment(
                                                        "another comment",
                                                    ),
                                                    59..78,
                                                ),
                                            ),
                                            59..78,
                                        ),
                                    ),
                                    59..78,
                                ),
                                (
                                    IfCondition(
                                        (
                                            InlineIfCondition(
                                                (
                                                    Eq(
                                                        (
                                                            Number(
                                                                (
                                                                    2.0,
                                                                    86..87,
                                                                ),
                                                            ),
                                                            86..87,
                                                        ),
                                                        (
                                                            Number(
                                                                (
                                                                    2.0,
                                                                    91..92,
                                                                ),
                                                            ),
                                                            91..92,
                                                        ),
                                                    ),
                                                    86..92,
                                                ),
                                                (
                                                    Echo(
                                                        (
                                                            "echo",
                                                            94..98,
                                                        ),
                                                        (
                                                            Text(
                                                                [
                                                                    (
                                                                        Text(
                                                                            (
                                                                                "y",
                                                                                100..101,
                                                                            ),
                                                                        ),
                                                                        100..101,
                                                                    ),
                                                                ],
                                                            ),
                                                            99..102,
                                                        ),
                                                    ),
                                                    94..102,
                                                ),
                                            ),
                                            86..102,
                                        ),
                                    ),
                                    86..102,
                                ),
                                (
                                    IfCondition(
                                        (
                                            Comment(
                                                (
                                                    Comment(
                                                        "another",
                                                    ),
                                                    111..122,
                                                ),
                                            ),
                                            111..122,
                                        ),
                                    ),
                                    111..122,
                                ),
                                (
                                    Else(
                                        (
                                            InlineElse(
                                                (
                                                    "else",
                                                    130..134,
                                                ),
                                                (
                                                    Echo(
                                                        (
                                                            "echo",
                                                            136..140,
                                                        ),
                                                        (
                                                            Text(
                                                                [
                                                                    (
                                                                        Text(
                                                                            (
                                                                                "z",
                                                                                142..143,
                                                                            ),
                                                                        ),
                                                                        142..143,
                                                                    ),
                                                                ],
                                                            ),
                                                            141..144,
                                                        ),
                                                    ),
                                                    136..144,
                                                ),
                                            ),
                                            130..144,
                                        ),
                                    ),
                                    130..144,
                                ),
                                (
                                    Comment(
                                        (
                                            Comment(
                                                "comment",
                                            ),
                                            145..156,
                                        ),
                                    ),
                                    145..156,
                                ),
                                (
                                    Comment(
                                        (
                                            Comment(
                                                "super comment",
                                            ),
                                            164..181,
                                        ),
                                    ),
                                    164..181,
                                ),
                                (
                                    Comment(
                                        (
                                            DocString(
                                                "doc comment",
                                            ),
                                            189..205,
                                        ),
                                    ),
                                    189..205,
                                ),
                            ],
                        ),
                        5..210,
                    ),
                ),
                5..210,
            ),
            (
                Statement(
                    (
                        IfCondition(
                            (
                                "if",
                                216..218,
                            ),
                            (
                                InlineIfCondition(
                                    (
                                        Ge(
                                            (
                                                Var(
                                                    (
                                                        "age",
                                                        219..222,
                                                    ),
                                                ),
                                                219..222,
                                            ),
                                            (
                                                Number(
                                                    (
                                                        16.0,
                                                        226..228,
                                                    ),
                                                ),
                                                226..228,
                                            ),
                                        ),
                                        219..228,
                                    ),
                                    (
                                        Echo(
                                            (
                                                "echo",
                                                230..234,
                                            ),
                                            (
                                                Text(
                                                    [
                                                        (
                                                            Text(
                                                                (
                                                                    "Welcome",
                                                                    236..243,
                                                                ),
                                                            ),
                                                            236..243,
                                                        ),
                                                    ],
                                                ),
                                                235..244,
                                            ),
                                        ),
                                        230..244,
                                    ),
                                ),
                                219..244,
                            ),
                            [
                                (
                                    Comment(
                                        "comment",
                                    ),
                                    245..256,
                                ),
                                (
                                    Comment(
                                        "comment in between",
                                    ),
                                    260..282,
                                ),
                            ],
                            Some(
                                (
                                    InlineElse(
                                        (
                                            "else",
                                            286..290,
                                        ),
                                        (
                                            Echo(
                                                (
                                                    "echo",
                                                    292..296,
                                                ),
                                                (
                                                    Text(
                                                        [
                                                            (
                                                                Text(
                                                                    (
                                                                        "Entry",
                                                                        298..303,
                                                                    ),
                                                                ),
                                                                298..303,
                                                            ),
                                                            (
                                                                Text(
                                                                    (
                                                                        "not",
                                                                        304..307,
                                                                    ),
                                                                ),
                                                                304..307,
                                                            ),
                                                            (
                                                                Text(
                                                                    (
                                                                        "allowed",
                                                                        308..315,
                                                                    ),
                                                                ),
                                                                308..315,
                                                            ),
                                                        ],
                                                    ),
                                                    297..316,
                                                ),
                                            ),
                                            292..316,
                                        ),
                                    ),
                                    286..316,
                                ),
                            ),
                        ),
                        216..316,
                    ),
                ),
                216..316,
            ),
            (
                Statement(
                    (
                        Comment(
                            (
                                Comment(
                                    "another comment",
                                ),
                                317..336,
                            ),
                        ),
                        317..336,
                    ),
                ),
                317..336,
            ),
        ],
    ),
    [],
)
